components/sup-client/src/lib.rs

```diff
@@ -146,27 +152,61 @@ impl SrvClient {
     ///
     /// Returns a stream of `SrvMessage`'s representing the server response.
     pub async fn request(
-        address: &ListenCtlAddr,
+        addr: Option<&ResolvedListenCtlAddr>,
         request: impl Into<SrvMessage> + fmt::Debug)
         -> Result<impl Stream<Item = Result<SrvMessage, io::Error>>, SrvClientError> {
-        let socket = TcpStream::connect(address.as_ref()).await?;
-        let mut socket = Framed::new(socket, SrvCodec::new());
+        let addr = Self::ctl_addr(addr)?;
+        let tcp_stream = TcpStream::connect(addr.addr()).await?;
+
+        // TLS configuration
+        let config = CliConfig::load()?;
+        let client_certificates = config.ctl_client_certificate
+                                        .map(CertificateChainCli::into_inner);
+        let client_key = config.ctl_client_key.map(PrivateKeyCli::into_inner);
+        let server_ca_certificates = config.ctl_server_ca_certificate
+                                           .map(RootCertificateStoreCli::into_inner);
+        let server_name_indication = config.ctl_server_name_indication;
+        let maybe_tls_config = if let Some(server_certificates) = server_ca_certificates {
+            let mut tls_config = TlsClientConfig::new();
+            tls_config.root_store = server_certificates;
+            if let Some(client_key) = client_key {
+                debug!("Configuring ctl-gateway TLS with client certificate");
+                tls_config.set_single_client_cert(client_certificates.unwrap_or_default(),
+                                                  client_key)?;
+            }
+            Some(Arc::new(tls_config))
+        } else {
+            None
+        };
```

Comment:
:thinking: Rather than having this function implicitly depend on `CliConfig`, it might be more clear to pass that in as an argument. 

Possibly better would be to give `CliConfig` a function that could generate the necessary TLS config object, and then pass _that_ into this function.

Reply:
:thinking: this is a difficult question; I wrestled with it as well. I actually made the opposite decision and removed the `secret` parameter from the `request` function. My reasoning was that whenever `request` was called we did the exact same thing: lookup the secret and then pass it to the function. This simply complicates the logic when we go to call `request`. Looking up the TLS parameters would be the same story. I agree it does not feel 100% correct to encapsulate this logic inside `request`. However, I do think we want it encapsulated somewhere. I am inclined to leave it in `request` for now and if we have a need for a more general `request` function we can make the change then. What do you think?

Reply:
Maybe we could split the difference... still depend implicitly on `CliConfig` within the function, but encapsulate the TLS parameter creation as a method on `CliConfig`. If that's not too gnarly, that seems like a more natural location for the logic than here (similar arguments might be made for the server TLS config). 

If that looks gross, though, we could leave it as is for the time being.

Reply:
Yep, that sounds like a good approach. Addressed in https://github.com/habitat-sh/habitat/pull/7923/commits/91a106eae99ed7f4a77f1d4da00cac9d0e0c476b